gguf-py/gguf/constants.py

```diff
@@ -79,33 +77,35 @@ class SSM:
         TIME_STEP_RANK = "{arch}.ssm.time_step_rank"
 
     class Tokenizer:
-        MODEL            = "tokenizer.ggml.model"
-        PRE              = "tokenizer.ggml.pre"
-        LIST             = "tokenizer.ggml.tokens"
-        TOKEN_TYPE       = "tokenizer.ggml.token_type"
-        TOKEN_TYPE_COUNT = "tokenizer.ggml.token_type_count"  # for BERT-style token types
-        SCORES           = "tokenizer.ggml.scores"
-        MERGES           = "tokenizer.ggml.merges"
-        BOS_ID           = "tokenizer.ggml.bos_token_id"
-        EOS_ID           = "tokenizer.ggml.eos_token_id"
-        UNK_ID           = "tokenizer.ggml.unknown_token_id"
-        SEP_ID           = "tokenizer.ggml.seperator_token_id"
-        PAD_ID           = "tokenizer.ggml.padding_token_id"
-        CLS_ID           = "tokenizer.ggml.cls_token_id"
-        MASK_ID          = "tokenizer.ggml.mask_token_id"
-        ADD_BOS          = "tokenizer.ggml.add_bos_token"
-        ADD_EOS          = "tokenizer.ggml.add_eos_token"
-        ADD_PREFIX       = "tokenizer.ggml.add_space_prefix"
-        HF_JSON          = "tokenizer.huggingface.json"
+        MODEL            = "tokenizer.model"  # STRING: e.g. llama, gpt2, etc...
+        TYPE             = "tokenizer.type"  # STRING: BPE, SPM, WPM, etc.
+        NORM             = "tokenizer.norm"  # OBJECT {"type": "ByteLevel", ...}
+        PRE              = "tokenizer.pre"  # OBJECT {"type": "ByteLevel", ...}
+        ADDED            = "tokenizer.added"  # ARRAY of OBJECTs: [{"id": 1, ...}, ...]
+        VOCAB            = "tokenizer.vocab"  # ARRAY of STRINGs: ["[BOS]", ...]
+        MERGES           = "tokenizer.merges"  # ARRAY of STRINGs: ["▁ t", ...]
+        TOKEN_TYPE       = "tokenizer.token_type"  # ARRAY of INT [2, ...]
+        TOKEN_TYPE_COUNT = "tokenizer.token_type_count"  # BERT token types
+        SCORES           = "tokenizer.scores"  # WPM only
+        BOS_ID           = "tokenizer.bos_token_id"
+        EOS_ID           = "tokenizer.eos_token_id"
+        UNK_ID           = "tokenizer.unknown_token_id"
+        SEP_ID           = "tokenizer.seperator_token_id"
+        PAD_ID           = "tokenizer.padding_token_id"
+        CLS_ID           = "tokenizer.cls_token_id"
+        MASK_ID          = "tokenizer.mask_token_id"
+        ADD_BOS          = "tokenizer.add_bos_token"
+        ADD_EOS          = "tokenizer.add_eos_token"
+        ADD_PREFIX       = "tokenizer.add_space_prefix"
```

Comment:
I need consensus on this. If anyone disagrees with this, I'm open to alternatives, but this will become more clear once I'm ready to full flesh everything out. I already looked at the C/C++ source code, but I'm waiting on @jaime-m-p and @mofosyne PRs cause it will cause conflicts otherwise. I published a "[tokenizers](https://github.com/teleprint-me/tokenizers)" repo on my account if you want to play around with the toy examples. I have it "semi-functional" there. I needed to separate it from this branch though so I could freely experiment with some ideas without radically changing anything here.

Reply:
> I published a "[tokenizers](https://github.com/teleprint-me/tokenizers)" repo on my account if you want to play around with the toy examples. I have it "semi-functional" there.

Interesting! I think there's likely a way to make your idea of the registry file work with the intentions of the existing GGUF metadata keys. It could even be included directly in the GGUF model files. Something similar to `tokenizer.huggingface.json`, but for a subset of it (to avoid including the vocab twice (or not?)), to include the pretokenizer, normalizer, and other relevant configs in the GGUF models.

From what I see in your `vocab.cpp` I think this is a great start! I didn't yet think about parsing JSON in C++. Sometimes I forget that `common/json.hpp` (aka `nlohmann/json`) is there.

I'm throwing some ideas here.

Maybe `convert-hf-to-gguf.py` should simply start including `tokenizer.json` (and `tokenizer_config.json`?) in the `tokenizer.huggingface.json` GGUF metadata key?

This way the tokenizer problems could be solved later without necessarily having to re-convert the model files. But I don't know to what extent parsing JSON (especially something like `Llama 3`'s 15MB `tokenizer.json`) would affect the load time. If it's too slow, maybe subsets of it could be in different GGUF metadata keys?

---

Even though I tend to nitpick a lot, it's nice to see that you are experimenting with this.

Reply:
Yeah, the refactored structure here is aligned with that idea. Partition the most common key-value pairs so we only load, read, and write once. This should negligibly affect the load time isolating what's needed at runtime.

I'm keeping in mind that we also rely on the SPM `tokenizer.model` files as well which is why I kept the rest "in-tact" for the most part. I'm thinking about it in terms of compatibility.

I'd also like to eventually build a BPE tokenizer for llama.cpp that we can use natively. I think this would be cool and potentially useful for training. This would effectively reduce the reliance on third-party sources for in-house models.

I'm very interested in training my own models which is why this fascinates me. I have a few older cards, so I'm thinking I can rig them up and use vulkan, but that's another task entirely.

I digress, but in short, I think I may partially revert the structure for backwards compatibility as I understand this is a valid concern.